Uploading error logs to GitHub from Jenkins CI Runs

TerrenceMcGuinness-NOAA opened the pull request:
# Description

This PR enhances the user experience within GitHub for when errors occur during build and running of CI cases from within the PR messages.  This is done by uploading the error logs to GitHub Gists and then publishing the links to them along with the full paths of the logs on disk.

This PR adds the python Class **GitHubPR** in `${HOMEgfs}/ci/scripts/utils/githubpr.py`  by inhearting the GitHub Class from **pyGitHub**.  We use this module to introduce a helper python utility that can publish a list of log files into a GitHub Gist and/or the designated branch **error_logs** in the **emcbot**  repo **ci-global-workflows** for storing error log files for review from any git configured terminal.  This upload feature will also create persistence of errors over time.

Also the `build_all.sh -k` script has been updated to support a "quick kill" feature (thanks David) that stops the parallel builds whenever one fails and creates an error_log file that has the paths to the error files that are also uploaded and published in the PR messages in GitHub.

# Type of change
- New feature (adds functionality)
- Maintenance (code refactor, adding a GitHub Class to use GItHub CLI using python)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
This PR has been run within Jenkins and has reported out the results [correctly to the messages](https://github.com/NOAA-EMC/global-workflow/pull/2422#issuecomment-2015436622).

TerrenceMcGuinness-NOAA committed:
trying status of CI as a workflow action

TerrenceMcGuinness-NOAA committed:
added PR number as input

TerrenceMcGuinness-NOAA committed:
Update orion_status.yaml

type labeled on new line

TerrenceMcGuinness-NOAA committed:
take out PR on on

TerrenceMcGuinness-NOAA committed:
take out PR on on

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops syntax until

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops syntax until singleton

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops syntax until set faulse

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops syntax until direct

TerrenceMcGuinness-NOAA committed:
take out PR on on do loops needed noop

TerrenceMcGuinness-NOAA committed:
take out PR jobs with no steps

TerrenceMcGuinness-NOAA committed:
take out PR jobs need steps

TerrenceMcGuinness-NOAA committed:
take out PR jobs need steps

TerrenceMcGuinness-NOAA committed:
added reusable workflow for getting lables for each job

TerrenceMcGuinness-NOAA committed:
fixed uses path

TerrenceMcGuinness-NOAA committed:
added dependances for main jobs and name in getlabels

TerrenceMcGuinness-NOAA committed:
change order in needs

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
took out poll

TerrenceMcGuinness-NOAA committed:
took out secrets required line

TerrenceMcGuinness-NOAA committed:
took out secrets required line

TerrenceMcGuinness-NOAA committed:
added {{}} on inputs.pr_number

TerrenceMcGuinness-NOAA committed:
fixed other pr_number assiginments

TerrenceMcGuinness-NOAA committed:
EOF error working

TerrenceMcGuinness-NOAA committed:
EOF error working II

TerrenceMcGuinness-NOAA committed:
ambiguous redrirect to GITHUB_OUTPUT

TerrenceMcGuinness-NOAA committed:
tryig to get needs output from get_Ready

TerrenceMcGuinness-NOAA committed:
still not getting outputs

TerrenceMcGuinness-NOAA committed:
quotes on GITHUB_OUTPUT

TerrenceMcGuinness-NOAA committed:
trying crazy Sys IO

TerrenceMcGuinness-NOAA committed:
addeing yml file

TerrenceMcGuinness-NOAA committed:
remove yml file

TerrenceMcGuinness-NOAA committed:
added () on DONE

TerrenceMcGuinness-NOAA committed:
adding env back to GITHUB_OUPUT

TerrenceMcGuinness-NOAA committed:
added read write permissions to reuse

TerrenceMcGuinness-NOAA committed:
fixed logic in status workflow for on set

TerrenceMcGuinness-NOAA committed:
noop on getlables

TerrenceMcGuinness-NOAA committed:
issue with orion status

TerrenceMcGuinness-NOAA committed:
misspelled yaml as yml

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into workflow_actions

Commit:
no op for self test

Commit:
no op for self test

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
adde pygithug into pr_list_database

TerrenceMcGuinness-NOAA committed:
updated newer changes to pr database script

TerrenceMcGuinness-NOAA committed:
fixed open ready to list of PRs

TerrenceMcGuinness-NOAA committed:
tweeked discription in update_database

TerrenceMcGuinness-NOAA committed:
pycode updated spaces

TerrenceMcGuinness-NOAA committed:
added some pycodestyle spacings

TerrenceMcGuinness-NOAA committed:
couple more pycodestyle spaces

TerrenceMcGuinness-NOAA committed:
put back open as requiremnt of PRs that labels are gotten from

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into pr_list_withPyGitHub

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
remove extra gitworkfow work stuck in my forked develop branch

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into pr_list_withPyGitHub

TerrenceMcGuinness-NOAA committed:
adde base git_pr_list and also git_ci_pr_list that optionally filters by host and defaulted by Ready lables to may be used for getting Kill list

TerrenceMcGuinness-NOAA committed:
added pycodestyle spaces

TerrenceMcGuinness-NOAA committed:
moved data base state code and GitHub module for class into utils directory as the are not scripts

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into pr_list_withPyGitHub

TerrenceMcGuinness-NOAA committed:
trying to push file using pyGitHub

TerrenceMcGuinness-NOAA committed:
Merge branch 'pr_list_withPyGitHub' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into pr_list_withPyGitHub

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into pr_list_withPyGitHub

TerrenceMcGuinness-NOAA committed:
tested and added string from subdir to publish_logs

TerrenceMcGuinness-NOAA committed:
added publish logs code with GitHubPR class module

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
added TOKEN option fo GitHubPR

TerrenceMcGuinness-NOAA committed:
added publish code in Jenkinsfile

TerrenceMcGuinness-NOAA committed:
pystyle spaces

TerrenceMcGuinness-NOAA committed:
pystyle spaces on operorator

TerrenceMcGuinness-NOAA committed:
updated discription in iniit

TerrenceMcGuinness-NOAA committed:
updated another space for operator

TerrenceMcGuinness-NOAA committed:
Update githubpr.py

r

TerrenceMcGuinness-NOAA committed:
misspelled scripts in path to log file code

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
made GitHubPR an inhearted class and added gist and repo as agruments for file in publishing a file

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature/upload_ci_logs' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
got all the paths right and gist and repo switches work

TerrenceMcGuinness-NOAA committed:
added doctring for inherated class with attributes

TerrenceMcGuinness-NOAA committed:
added build all with k feature

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature/upload_ci_logs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
updated publish logs code to handle lists of files

TerrenceMcGuinness-NOAA committed:
correctly loop over list of files as read file objects already

TerrenceMcGuinness-NOAA committed:
folded in list basse gist and reop error log published script into Jenkins file with messaging

TerrenceMcGuinness-NOAA committed:
shell norm quotes on build all

TerrenceMcGuinness-NOAA committed:
more shell norm quotes on build all

TerrenceMcGuinness-NOAA committed:
bracket issues

TerrenceMcGuinness-NOAA committed:
still had bracket issues

TerrenceMcGuinness-NOAA committed:
still had bracket issues and removed he HOMEgfs workspace

TerrenceMcGuinness-NOAA committed:
nested error varibles fixed

TerrenceMcGuinness-NOAA committed:
updated publish files script name in Jenkins

TerrenceMcGuinness-NOAA committed:
test for empty log files and put in a delay to make sure they get populated

TerrenceMcGuinness-NOAA committed:
added k switch to build paramters

TerrenceMcGuinness-NOAA committed:
need double quotes to resolve env vars

TerrenceMcGuinness-NOAA committed:
extra )

TerrenceMcGuinness-NOAA committed:
need double quotes to resolve env vars

TerrenceMcGuinness-NOAA committed:
take out sleep

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
syntax error in checkout if main error log file was present and non-zero

TerrenceMcGuinness-NOAA committed:
still had anogher syntax error in checkout if main error log file was present and non-zero

TerrenceMcGuinness-NOAA committed:
needed base name for gist file name

TerrenceMcGuinness-NOAA committed:
fixed file names in git as list and added delete funciton

TerrenceMcGuinness-NOAA committed:
added some white spaces for readablity

TerrenceMcGuinness-NOAA committed:
pynorm spaces in publish code

TerrenceMcGuinness-NOAA committed:
had need varible to narrowly scoped

TerrenceMcGuinness-NOAA committed:
still got scope of error_logs_message to tight

TerrenceMcGuinness-NOAA committed:
comment if no error logs are found when cases fail

TerrenceMcGuinness-NOAA committed:
fixing already exits issue

TerrenceMcGuinness-NOAA committed:
trying to fix if exixts

TerrenceMcGuinness-NOAA committed:
still trying

TerrenceMcGuinness-NOAA committed:
still try try

TerrenceMcGuinness-NOAA committed:
adde try for when PR lead header is already in repo

TerrenceMcGuinness-NOAA committed:
adde try for when PR lead header is already in repo

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
updating devlop to match authoritative repo

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/upload_ci_logs

DavidHuber-NOAA committed:
Fixed iterators in build_all.

TerrenceMcGuinness-NOAA committed:
added comments and added markdown to GitHumb messaging for error logs

TerrenceMcGuinness-NOAA committed:
added extra space in message and a comment on top for workspaces

TerrenceMcGuinness-NOAA committed:
fixed and tested path_header update when exisits in error_log repo

TerrenceMcGuinness-NOAA committed:
Merge pull request #265 from DavidHuber-NOAA/fix/build_iters

Fixed iterators in build_all

TerrenceMcGuinness-NOAA committed:
added (to view contents of these file) to make it cleaer what the link dones

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature/upload_ci_logs' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
added (to view contents of these file) to make it cleaer what the link dones again

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
fixed conflicts with Jenknins file

TerrenceMcGuinness-NOAA committed:
fixec pycode style splaces

WalterKolczynski-NOAA commented on ci/scripts/utils/githubpr.py:
```diff
@@ -0,0 +1,125 @@
+        if repo_url is None:
+            repo_url = os.environ.get("REPO_URL")
+        match = re.search(r"github\.com/(.+)", repo_url)
+        repo_identifier = match.group(1)[:-4]
```
Why just the first four characters?

TerrenceMcGuinness-NOAA replied:
Completely arbitrary.  Intend to be update to increment numerally in next iteration.

TerrenceMcGuinness-NOAA replied:
It was to remove the `.com` portion after the match as it converts from full path URL to **GitHub**'s user/repo format.

WalterKolczynski-NOAA replied:
Sorry, you're right. My manual test of slicing wasn't great.

WalterKolczynski-NOAA commented on ci/scripts/utils/githubpr.py:
```diff
@@ -0,0 +1,125 @@
+        pr_list = []
+        for pull in self.pulls:
+            pr_list.append(pull.number)
+        return pr_list
```
This can be one line using list notation (does it even need to be a method?)
```suggestion
        return [ pull.number for pull in self.pulls ]
```

WalterKolczynski-NOAA commented on ci/scripts/utils/githubpr.py:
```diff
@@ -0,0 +1,125 @@
+                if state in label.name:
+                    if host is not None:
+                        if host.lower() in label.name.lower():
+                            pr_list.append(pull.number)
```
```suggestion
                            pr_list.append(pull.number)
                            break
```

WalterKolczynski-NOAA commented on ci/scripts/utils/githubpr.py:
```diff
@@ -0,0 +1,125 @@
+                        if host.lower() in label.name.lower():
+                            pr_list.append(pull.number)
+                    else:
+                        pr_list.append(pull.number)
```
```suggestion
                        pr_list.append(pull.number)
                        break
```

TerrenceMcGuinness-NOAA replied:
How does this help?

WalterKolczynski-NOAA replied:
Otherwise you continue looping over labels until you run out of them (and may even add a PR to the list twice).

TerrenceMcGuinness-NOAA replied:
The list results needs to be exhaustive.  We need to check all PRs and continue to find ones that match such labels.

WalterKolczynski-NOAA replied:
This should only break the innermost loop, which is over labels, not the PRs, and the PR currently being looked at is already added.

TerrenceMcGuinness-NOAA replied:
oh ok, let me take a closer look.

TerrenceMcGuinness-NOAA replied:
Oh I see, host is fixed, so its first occurrence is sufficient, and Kill would also be a unique criteria once - got it.

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+from argparse import ArgumentParser, ArgumentTypeError, FileType
+
+
+def is_file(path):
```
This function name doesn't really match what it is doing. `is_file` sound like it should return a bool about whether the file exists (as `os.path.isfile` does). It shouldn't be raising an exception if the file doesn't exist.

WalterKolczynski-NOAA replied:
Is this even being used?

TerrenceMcGuinness-NOAA replied:
Oh right, it was OBE when I updated the code to allow --file to be list and used FileType instead.  I'll remove it.

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+    return path
+
+
+def input_args():
```
```suggestion
def parse_args():
```

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+    parser.add_argument('--file', help='path to file for uploading to GitHub', required=False, type=FileType('r'), nargs='+')
+    parser.add_argument('--gist', help='create a gist of the file', nargs=1, metavar='identifier_string', required=False)
+    parser.add_argument('--repo', help='create a file in a repo', nargs=1, metavar='path_header', required=False)
+    parser.add_argument('--delete_gists', help='deletes all the gits from authenticated user', action='store_true', default=False, required=False)
```
Delete *all* of them?

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+    return args
+
+
+def random_string() -> str:
+    characters = string.ascii_letters + string.digits + string.punctuation
+    return re.sub('[^A-Za-z0-9 ]+', '', ''.join(secrets.choice(characters) for _ in range(6)))
```
I don't think you need `re` here. Also, is there a reason it needs to be use secrets?

```suggestion
def random_string(length=6) -> str:
    characters = string.ascii_letters + string.digits + string.punctuation
    return ''.join(secrets.choice(characters) for _ in range(length))
```

or

```suggestion
def random_string(length=6) -> str:
    characters = string.ascii_letters + string.digits + string.punctuation
    return ''.join(random.choices(characters, k=length))
```

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+    emcbot_ci_url = "https://github.com/emcbot/ci-global-workflows.git"
+    emcbot_gh = GitHubPR(repo_url=emcbot_ci_url)
+
+    if args.gist:  # Add error logs to a gist in GitHub emcbot's account
```
Each of these ifs should be a function.

TerrenceMcGuinness-NOAA replied:
oh I suppose.

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+    """
+    parser = ArgumentParser(description=description)
+
+    parser.add_argument('--file', help='path to file for uploading to GitHub', required=False, type=FileType('r'), nargs='+')
+    parser.add_argument('--gist', help='create a gist of the file', nargs=1, metavar='identifier_string', required=False)
+    parser.add_argument('--repo', help='create a file in a repo', nargs=1, metavar='path_header', required=False)
+    parser.add_argument('--delete_gists', help='deletes all the gits from authenticated user', action='store_true', default=False, required=False)
```
Are these mutually exclusive? Should this be modal like setup_exp?

TerrenceMcGuinness-NOAA replied:
They are not.  You can use and/or with `--gist` and `--repo`.

TerrenceMcGuinness-NOAA replied:
I think it's a good idea to only use one though and update this too.

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -0,0 +1,107 @@
+        gist = emcbot_gh.user.create_gist(public=True, files=gist_files, description=f"error log file from CI run {args.gist[0]}")
+        print(gist.html_url)
+
+    if args.repo:  # Upload error logs to emcbot's ci-global-workflows error_logs branch
```
Are we even going to use this option? I thought we had settled on gists.

TerrenceMcGuinness-NOAA replied:
Yes I am going to it for persistence.  We can remove from the reporting if you like.

WalterKolczynski-NOAA replied:
Okay, that's fine.

WalterKolczynski-NOAA commented on sorc/build_all.sh:
```diff
@@ -228,11 +257,31 @@ while [[ ${builds_started} -lt ${#build_jobs[@]} ]]; do
+   if [[ "${_quick_kill}" == "YES" ]]; then
+      check_builds
+      build_stat=$?
+      if [[ ${build_stat} != 0 ]]; then
```
```suggestion
      if (( build_stat != 0 )); then
```

WalterKolczynski-NOAA commented on sorc/build_all.sh:
```diff
@@ -193,6 +197,31 @@ fi
 procs_in_use=0
 declare -A build_ids
 
+check_builds()
```
```suggestion
kill_if_failure()
```

WalterKolczynski-NOAA commented on sorc/build_all.sh:
```diff
@@ -228,11 +257,31 @@ while [[ ${builds_started} -lt ${#build_jobs[@]} ]]; do
-while [[ ${#build_jobs[@]} -gt 0 ]]; do
+while [[ "${#build_jobs[@]}" -gt 0 ]]; do
+
+   # If requested, check if any build has failed and exit if so
```
I'm not sure we need this in both places. @DavidHuber-NOAA can you take a look at this script's changes?

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/githubpr.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update sorc/build_all.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
put in check to make sure at least  gitst or file is used

TerrenceMcGuinness-NOAA committed:
and in arg logic and python style spaces

TerrenceMcGuinness-NOAA committed:
removed is_file as it was replaced with FileType

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/publish_logs.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
added suggestions with args

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/publish_logs.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
made 4 random char function better from walt

TerrenceMcGuinness-NOAA committed:
indentation wrong

TerrenceMcGuinness-NOAA committed:
try inc over rnd

TerrenceMcGuinness-NOAA committed:
replaced rnd chars with an incrementing tag

TerrenceMcGuinness-NOAA committed:
updated arg tests

TerrenceMcGuinness-NOAA committed:
pynorms space

TerrenceMcGuinness-NOAA committed:
had an extra bracket in Jenkinsfile

TerrenceMcGuinness-NOAA committed:
had to redo Jenkinsfile merge conflics

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of https://github.com/TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
not both needed args was supposed to have an and logical operator between them

TerrenceMcGuinness-NOAA commented:
Manually remove Build Label because messaging to **GitHub** and label updates failed due to _Job's SCM is not GitHub_ error.

TerrenceMcGuinness-NOAA commented:
@WalterKolczynski-NOAA The **Run Tests** _Failed to build stage_ was a forced error I administered to test the error files being [published](https://gist.github.com/emcbot/23d6f114ad4b8b69b2e72ac415a796d4).

I added the gist link here by hand because Jenkins had one of those authentication errors from the Pipeline GitHub SCM Plugin.

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/upload_ci_logs

TerrenceMcGuinness-NOAA committed:
moved arg conditionals into functions and added docstrings to publishing logs code

TerrenceMcGuinness-NOAA committed:
fixe correct path in repo pub

TerrenceMcGuinness-NOAA committed:
updated representations of repo branch and path for pub log file

TerrenceMcGuinness-NOAA committed:
pynorms white space removal

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/githubpr.py

Skipping inner look first occurrence works because host is fixed (i.e. this code is run on a host-by-host basis).

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/githubpr.py

We can short circuit this inner loop because for a given PR the search key would be unique.

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

WalterKolczynski-NOAA commented on ci/scripts/utils/publish_logs.py:
```diff
@@ -50,58 +22,89 @@ def input_args():
     args = parser.parse_args()
-    if not args.delete_gists and not args.file:
-        parser.error("--file is required when --delete_gists is not used")
+    # if not args.delete_gists and not args.file:
+    #    parser.error("--file is required when --delete_gists is not used")
+    if not args.gist and not args.repo:  # At least one of the two is required
+        parser.error("--gist or --repo is required")
+    if args.gist and args.repo:  # can not use booth
+        parser.error("only use --gist or --repo")
```
Use `!=`?
```suggestion
    if bool(args.gist) != bool(args.repo):  # Exactly one of the two is required
        parser.error("Exactly one of --gist and --repo is required")
```

WalterKolczynski-NOAA replied:
Is there a reason not to allow both?

TerrenceMcGuinness-NOAA replied:
clever

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/utils/publish_logs.py

clever

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>
